### PR TITLE
Reconnect host on provider add

### PR DESCRIPTION
**Pull request by pkliczewski**
We need to make sure that when we remove a provider and add it again we
will keep using the same host entities. Without this change add and remove
provider cause to duplicate hosts.

Fixes https://bugzilla.redhat.com/1528859

### CODE REVIEW

**Comment by pkliczewski** [comment]
[Bug 1528859 - [RHV][Graph Refresh] "Archived" Hosts are displayed several times on CFME, with Power state: ON ("Running")](https://trello.com/c/lKLRbQNt/286-bug-1528859-rhvgraph-refresh-archived-hosts-are-displayed-several-times-on-cfme-with-power-state-on-running)

**Comment by pkliczewski** [comment]
@miq-bot assign @Ladas @agrare

**Comment by pkliczewski** [comment]
@miq-bot assign @Ladas

**Comment by pkliczewski** [comment]
@agrare please review

**Comment by pkliczewski** [comment]
@miq-bot add_label gaprindashvili/yes

File: app/models/manageiq/providers/redhat/inventory_collection_default/infra_manager.rb [hunk 1]
```diff
@@ -76,6 +76,39 @@ def root_folders(extra_attributes = {})
       attributes.merge!(extra_attributes)
     end
 
+    def hosts(extra_attributes = {})
+      reconnect_block = lambda do |inventory_collection, inventory_objects_index, attributes_index|
```
Comment (Ladas) [inline app/models/manageiq/providers/redhat/inventory_collection_default/infra_manager.rb]: we can use the same code for Vms, so we are compatible with old refresh, @agrare what else do we reconnect?

Comment (agrare) [reply app/models/manageiq/providers/redhat/inventory_collection_default/infra_manager.rb]: Great idea, just hosts and vms currently.

**Comment by agrare** [comment]
@Ladas @pkliczewski this probably belongs in the main [`ManagerRefresh::InventoryCollectionDefault::InfraManager`](https://github.com/ManageIQ/manageiq/blob/master/app/models/manager_refresh/inventory_collection_default/infra_manager.rb) since it isn't specific to ovirt

**Comment by pkliczewski** [comment]
@agrare will move it to the main repo.

**Comment by Ladas** [comment]
@agrare right, so @pkliczewski is adding it here because e.g. VmWare reconnect is different. In here we will reconnect using only ems_ref.

**Comment by agrare** [comment]
@Ladas you mean for hosts?  That is different behavior between normal and graph refresh then.

**Comment by Ladas** [comment]
@agrare right @pkliczewski says they need just reconnecting by ```:ems_ref```, because they don't have the scenario of loosing the whole RHEV DB, that would lead to new ems_refs , like VmWare

**Comment by agrare** [comment]
@Ladas that's not the only reason we lookup by hostname, if you drop a host from a provider and connect it to another this would also pick that up.

**Comment by pkliczewski** [comment]
Specs for RHV available here https://github.com/ManageIQ/manageiq/pull/16750

File: spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_async_graph_spec.rb [hunk 2]
```diff
@@ -48,6 +48,40 @@ def load_response_mock_for(filename)
...
+                             :ems_ref_obj           => "/api/vms/3a9401a0-bf3d-4496-8acf-edd3e903511f",
+                             :uid_ems               => "3a9401a0-bf3d-4496-8acf-edd3e903511f",
+                             :vendor                => "redhat",
+                             :raw_power_state       => "up",
+                             :location              => "3a9401a0-bf3d-4496-8acf-edd3e903511f.ovf")
+
+    EmsRefresh.refresh(@ems)
+    @ems.reload
+
+    vm = VmOrTemplate.find_by(:uid_ems => "3a9401a0-bf3d-4496-8acf-edd3e903511f")
```
Comment (Ladas) [inline spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_async_graph_spec.rb]: probably also assert that 

```
expect(VmOrTemplate.where(:uid_ems => "3a9401a0-bf3d-4496-8acf-edd3e903511f").count).to eq 1
```

To make sure the Vm is not duplicated ,same for the Host

Comment (pkliczewski) [reply spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_async_graph_spec.rb]: OK

Comment (pkliczewski) [reply spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_async_graph_spec.rb]: Done, @Ladas are you ok with the specs now?

Comment (Ladas) [reply spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_async_graph_spec.rb]: yes, awesome

**Comment by pkliczewski** [comment]
@masayag @oourfali please review

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit bc5bd40f2d640fa8a62154f05bd8035c88c2c7e1
Author: Moti Asayag <masayag@redhat.com>
Date:   Wed Jan 10 10:32:48 2018 +0200

    Merge pull request #189 from pkliczewski/reconnect
    
    Reconnect host on provider add
    (cherry picked from commit 057e7d051f62256e488dea83ab004714880755aa)
    
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1532332
```

**Pull request merged**